File: backend/session/paths.go
package session

import (
	"os"
	"path/filepath"
	"strings"
)

const (
	AgentDirName    = ".enough"
	AgentSubdir     = "agent"
	SessionsSubdir  = "sessions"
	CurrentVersion  = 1
)

func HomeAgentDir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, AgentDirName, AgentSubdir), nil
}

func EncodeCWD(cwd string) string {
	s := strings.TrimPrefix(cwd, string(os.PathSeparator))
	if vol := filepath.VolumeName(cwd); vol != "" {
		s = strings.TrimPrefix(cwd, vol)
		s = strings.TrimPrefix(s, string(os.PathSeparator))
	}
	s = strings.NewReplacer(
		string(os.PathSeparator), "-",
		":", "-",
	).Replace(s)
	return "--" + s + "--"
}

func SessionDir(cwd string) (string, error) {
	agentDir, err := HomeAgentDir()
	if err != nil {
		return "", err
	}
	dir := filepath.Join(agentDir, SessionsSubdir, EncodeCWD(cwd))
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return "", err
	}
	return dir, nil
}

File: backend/session/paths.ts
// PORT: backend/session/paths.go

import { Effect } from "effect";
import fs from "node:fs/promises";
import path from "node:path";
import { home_dir } from "../hollowhome/home";

export const agent_dir_name = ".hollow";
export const agent_subdir = "agent";
export const sessions_subdir = "sessions";
export const current_version = 1;

export const home_agent_dir = (): Effect.Effect<string, Error> => {
  return Effect.succeed(path.join(home_dir(), agent_subdir));
};

export const encode_cwd = (cwd: string): string => {
  let s = cwd;
  // Strip leading path separator
  if (s.startsWith(path.sep)) {
    s = s.substring(path.sep.length);
  }
  // Handle Windows volume name (e.g. C:)
  const drive_match = s.match(/^[a-zA-Z]:/);
  if (drive_match) {
    const vol = drive_match[0];
    s = s.substring(vol.length);
    if (s.startsWith(path.sep)) {
      s = s.substring(path.sep.length);
    }
  }
  // Replace separators and colons with "-"
  s = s.split(path.sep).join("-").split(":").join("-");
  return "--" + s + "--";
};

export const session_dir = (cwd: string): Effect.Effect<string, Error> => {
  return home_agent_dir().pipe(
    Effect.flatMap((agent_dir) =>
      Effect.tryPromise({
        try: async () => {
          const dir = path.join(agent_dir, sessions_subdir, encode_cwd(cwd));
          await fs.mkdir(dir, { recursive: true, mode: 0o700 });
          return dir;
        },
        catch: (cause) => cause instanceof Error ? cause : new Error(String(cause)),
      }),
    ),
  );
};

/*
PORT STATUS
source path: backend/session/paths.go
source lines: 46
draft lines: 51
confidence: high
status: phase_b_compile
*/
